pkg/controller/infinispan/encryption.go

```diff
@@ -0,0 +1,177 @@
+			}
+			return nil
+		}
+	}
+
+	secretContains := func(keys ...string) bool {
+		for _, k := range keys {
+			if _, ok := tlsSecret.Data[k]; !ok {
+				return false
+			}
+		}
+		return true
+	}
+
+	if secretContains(EncryptKeystoreName) {
+		// If user provide a keystore in secret then use it ...
+		c.Keystore.Path = fmt.Sprintf("%s/%s", EncryptMountPath, EncryptKeystoreName)
+		c.Keystore.Password = string(tlsSecret.Data["password"])
+		c.Keystore.Alias = string(tlsSecret.Data["alias"])
+	} else if secretContains("tls.key", "tls.crt") {
+		configureNewKeystore(c)
+	}
+
+	if i.IsClientCertEnabled() {
+		caPem := tlsSecret.Data[EncryptClientCAName]
+		return configureClientCert(caPem, i, c, tlsSecret, client)
+	}
+	return nil
+}
+
+func configureClientCert(caPem []byte, m *v1.Infinispan, c *config.InfinispanConfiguration, tlsSecret *corev1.Secret, client client.Client) error {
+	// If the user explicitly provides a truststore use that, otherwise look for individual client certificates and create one
+	c.Truststore.Path = fmt.Sprintf("%s/%s", EncryptMountPath, EncryptTruststoreName)
+	c.Endpoints.ClientCert = string(m.Spec.Security.EndpointEncryption.ClientCert)
+
+	password, passswordProvided := tlsSecret.Data[EncryptTruststorePasswordKey]
+
+	// If secret already contains a truststore only configure the password
+	if _, ok := tlsSecret.Data[EncryptTruststoreName]; ok {
+		if !passswordProvided {
+			return fmt.Errorf("The '%s' key must be provided when configuring an existing Truststore", EncryptTruststorePasswordKey)
+		}
+		c.Truststore.Password = string(password)
+		return nil
+	}
+
+	if !passswordProvided {
+		password = []byte(EncryptTruststorePassword)
+	}
+
+	certs := [][]byte{caPem}
+	for certKey := range tlsSecret.Data {
+		if strings.HasPrefix(certKey, EncryptClientCertPrefix) {
+			certs = append(certs, tlsSecret.Data[certKey])
+		}
+	}
+
+	truststore, err := generateTruststore(certs, string(password))
+	if err != nil {
+		return err
+	}
+
+	_, err = kube.CreateOrPatch(ctx, client, tlsSecret, func() error {
+		if tlsSecret.CreationTimestamp.IsZero() {
+			return errors.NewNotFound(corev1.Resource("secret"), m.GetEncryptionSecretName())
+		}
+		tlsSecret.Data[EncryptTruststoreName] = truststore
+		tlsSecret.Data[EncryptTruststorePasswordKey] = []byte(password)
+		return nil
+	})
+	return err
+}
+
+func generateTruststore(pemFiles [][]byte, password string) ([]byte, error) {
+	certs := []*x509.Certificate{}
+	for _, pemFile := range pemFiles {
+		pemRaw := pemFile
+		for {
+			block, rest := pem.Decode(pemRaw)
+			if block == nil {
+				break
+			}
+
+			if block.Type == "CERTIFICATE" {
+				cert, err := x509.ParseCertificate(block.Bytes)
+				if err != nil {
+					return nil, fmt.Errorf("Unable to parse certificate: %w", err)
+				}
+				certs = append(certs, cert)
+			} else {
+				log.Info(fmt.Sprintf("Ingoring pem entry type %s when generating truststore. Only CERTIFICATE is supported.", block.Type))
```

Comment:
We should send notifications like this to the console event log also when #346 will be developed